### ISSUE TITLE
Grid: Multiple Selection behaves strange in combination with Virtual Scrolling

### ISSUE THREAD

**Issue by paoluptecoi** [telerik/kendo-angular#889]
## I'm submitting a...
* Bug report 

## Current behavior I
When I select the first item and then scroll down to the last item hold SHIFT and click on the last item, then only those items currently visible and the first item will be selected.

## Expected behavior I
I would expect all items to be selected.

## Current behavior II
When I select the first item and then scroll down to the last item and click on the last item, then the last item and the first item will be selected. (The invisible first row was not deselected.)

## Expected behavior II
I would expect only the last item to be selected. Not the first item!

## Minimal reproduction of the problem with instructions

http://plnkr.co/edit/dcVwa6kX47Kuw0GZsqQP?p=preview

To reproduce I:
* select first row
* scroll to end
* hold SHIFT and select last row
* scroll up a little to see the problem

To reproduce II:
* reload
* select first row
* scroll to end
* click on last row
* scroll to first row to see the problem

## What is the motivation or use case for changing the behavior?
The selection of the rows should work like selection of rows always work: 
* Keep previously select with Ctrl
* Clear previously selected without Ctrl
* select from previously selected to currently selected with Shift

## Environment

Package versions:
- reproducible in Plunker

Browser:
- Chrome (desktop) version 60.0.3112.113 (Offizieller Build) (64-Bit)
- Firefox version 55.0.3 (32-Bit)

System:
- TypeScript version: 2.4.2
- Node version: v8.4.0 5.4.1
- Platform:  Windows

## A guess:
I think the problem is that the selection event handles only the rows of the currently loaded virtual page. It would be necessary to load all the rows.

**Comment by Raisolution** [comment]
Hi @paoluptecoi,

The described behavior is expected, as indeed the selection operates only over the currently visible items. We are currently working on a "select all" feature and improvements for similar scenarios.

**Comment by Raisolution** [comment]
The `select all` feature has been introduced in v1.5.0.

**Comment by paoluptecoi** [comment]
This issue was not resolved by the select all feature. Please note that mentioning the first and the last item was only to provide a simple expamle. Please use the second and the second to last item in my description and you will see that the problem is not resolved with a select all feature.

**Comment by Raisolution** [comment]
Yes, as mentioned the selection operates only over the currently rendered items, so it is not aware of the items in-between the second and the second to last.

**Comment by paoluptecoi** [comment]
Then please reopen the issue. This is a bug.

**Comment by Raisolution** [comment]
Hi @paoluptecoi,

The described behavior is by design - the selection operates only over the currently rendered items - in the current case the items from the last page (after the Grid was scrolled). We will update the documentation regarding the selection, in order to clarify this better.

**Comment by noariki** [comment]
To provide more context, in IE you are currently stuck with virtual scrolling, because of how bad performance with kendo grid is otherwise.
If it's by design then it's a **very bad design**.

**Comment by tsvetomir** [comment]
@noariki we're investigating the performance issues in IE, but they're not directly related.

To summarise the problem so far - the `kendoGridSelectBy` directive cannot select ranges across pages, as it does not have access to the full list of data items.

In theory, it could be made to work for local data by operating on the processed (sorted, filtered and grouped), but non-paged data collection of the grid. There are no plans to implement such workaround at the moment, as it won't function at all with remote data.

An alternative would be to persist the selection by row position. This would work for the scenario at hand, but will break whenever the data set has changed - including sorting, filtering and grouping changes. You can try [this prototype](https://stackblitz.com/edit/kendo-angular-grid-select-by-index?file=app%2Fselect-by-index.directive.ts) to get an idea for the limitations of this approach.

**Comment by noariki** [comment]
@tsvetomir Hi, thanks for the reply. I know that these issues are not connected, I was just pointing out that it's not some obscure scenario, people need this to work.
As for the prototype, at least in my application, I don't care if the selection is gone when sorting, filtering or grouping occurs. In fact, I want selection to be cleared in these cases. So I will try to use it as a work-around.

**Comment by richarddzh** [comment]
Please reopen this bug.
It does not make sense when user scrolls and click on a row without holding the ctrl button, the new row is added to the previous selection.
This shouldn't be the expected behavior.
Users should always expect the same behavior regardless of whether virtual scroll is enabled.
This is a bug.
You can't expect that the end user accept this bug as 'by design' by knowing that virtual scrolling is enabled. Virtual scrolling is implementation details that user should not be aware of.

**Comment by EoinGriffin-AI** [comment]
I am bemused by how this bug has been disregarded by the kendo team. It is a bad, flawed design.
It is a bug, and it still occurs.

It can be made to work for local data if the grid would return at least an array of selected row indices.
Such an array of selected indices would also be valuable for remote data too. An understandable limitation might be to _only_ return row indices (and not unloaded data) if it's using remote data, but that's still better than it is now.

**Comment by santoro-mariano** [comment]
Please, reopen this bug.
I'm seriously shocked with the telerik team answer, I can't believe this can be considered a "by design" behavior.
At a minimun, if I'm not holding the ctrl key when I'm clicking the row, it should not keep the previous row selected.

**Comment by tsvetomir** [comment]
We plan to bring the Grid selection in line with the [TreeList implementation](https://www.telerik.com/kendo-angular-ui/components/treelist/selection/#toc-persisting-the-selection). It moves the selection state outside of the component and allows more fine-grained control over it. See [runnable demo](https://stackblitz.com/edit/angular-akcnjt?file=app%2Fapp.component.ts).

We're still figuring out how to do so with minimal or no breaking changes, but we hope to have something to show for the  release in May or soon thereafter.

**Comment by EoinGriffin-AI** [comment]
@tsvetomir In that demo shift-selection is still buggy, but I'm happy to hear the issues are getting addressed.
Thank you for that!

For clarification about what I mean by buggy, if you select the first item, scroll to the bottom and then select the last item, then every item should be selected.
What happens now is that only the rendered items are selected. If you scroll up you'll see that nothing else is selected.

**Comment by tsvetomir** [comment]
@EoinGriffin-AI this is expected as the Grid doesn't have access to the items from other pages when using paging or virtualization.

Still, we would be able to handle this case by using the [TreeList-style custom selection](https://www.telerik.com/kendo-angular-ui/components/treelist/selection/custom-selection/).  If the data is available in memory, we can store the keys directly. If not, we can make a call to a service to obtain them. We could even store the selection as a range, given a fixed sort order.

**Comment by mbechev** [comment]
Reported in private support threads:
**1600138**
**1603950**

**Comment by jcorrea97** [comment]
@dtopalov I used your example and changed the pagesize, and when I select item 1, 3, 4 and 6 (all items using "ctrl") the event returns correctly, but if I scroll down and select an item by pressing "shift" the grid will deselect items 1,3 and 4 and the event does not return items unchecked in the event.
And I need to get this information because I need to get the rows that are not selected. How can I get the items that have been deselected?

I put a gif with the example:

https://cdn.discordapp.com/attachments/1215659099467808849/1275169938834849864/kendofu.gif?ex=66c4ea0c&is=66c3988c&hm=e1c9a362fe115d0e2e8eed242c6f7002f71e770729412d5e5eaaafa80ea9dbdd&
